java/src/main/native/src/RmmJni.cpp

```diff
@@ -349,19 +369,50 @@ class java_event_handler_memory_resource final : public device_memory_resource {
                                  on_dealloc_threshold_method, "onDeallocThreshold", total_after);
   }
 
-  bool supports_get_mem_info() const noexcept override { return resource->supports_get_mem_info(); }
+  void on_allocated_callback(std::size_t num_bytes, rmm::cuda_stream_view stream) {
+    JNIEnv *env = cudf::jni::get_jni_env(jvm);
+    env->CallVoidMethod(handler_obj, on_allocated_method, num_bytes);
+    if (env->ExceptionCheck()) {
+      throw std::runtime_error("onAllocated handler threw an exception");
+    }
+  }
 
-  std::pair<size_t, size_t> do_get_mem_info(rmm::cuda_stream_view stream) const override {
-    return resource->get_mem_info(stream);
+  void on_deallocated_callback(void *p, std::size_t size, rmm::cuda_stream_view stream) {
+    JNIEnv *env = cudf::jni::get_jni_env(jvm);
+    env->CallVoidMethod(handler_obj, on_deallocated_method, size);
+    if (env->ExceptionCheck()) {
+      env->ExceptionDescribe();
+      throw std::runtime_error("onDeallocated handler threw an exception");
```

Comment:
Not sure it's safe to throw here, as this is likely called within a destructor. Throwing in a destructor will likely cause the C++ runtime to terminate the application harshly.

Reply:
Sounds good. I won't call `ExceptionCheck` since there's nothing for me to handle. I expect the java code would see an exception after returning from the java-side of the JNI code and then throw, but it wouldn't throw within cuDF.

Reply:
Yeah confirmed we get an exception on the java side when not handled here, which seems fine.